### DISCUSSION TITLE
Double diffusion simulation

### DISCUSSION THREAD

**Discussion by LijoDXL** [CliMA/Oceananigans.jl#1302] [Computational science!] [answered]
First of all thank for creating an easy and the same time powerful modeling tool. I am trying to simulate a double diffusion criteria and study how mixing evolved under such conditions. As far as I understand, Turbulence Closure scheme implemented here has **same diffusivity coefficients for temperature and salinity**. Since double diffusion needs those coefficients to be different, could you **please point me to where I can set those values?** Also, how can I **feed  the observed temperature and salinity fields (available in NetCDF) as input to the simulation**. I am able to create a synthetic T(z) and S(z) profiles using the built-in model.grid and random noise but note sure how to set initial T(z) and S(z) from a NetCDF file

**Comment by ali-ramadhan** [comment, marked as answer]
Hi @LijoDXL! Glad you're finding Oceananigans easy to use.

> Since double diffusion needs those coefficients to be different, could you please point me to where I can set those values

You can set different diffusivities for different tracers, but I just realized that this feature might still be undocumented. You can pass a named tuple for when specifying some closures. For example:

```julia
julia> using Oceananigans

julia> grid = RegularCartesianGrid(topology=(Periodic, Periodic, Bounded), size=(16, 16, 16), extent=(1, 1, 1));

julia> model = IncompressibleModel(grid=grid, closure=IsotropicDiffusivity(ν=5, κ=(T=5, S=12)));

julia> model.closure.κ
(T = 5, S = 12)
```

> how can I feed the observed temperature and salinity fields (available in NetCDF) as input to the simulation.

You can read the observed temperature and salinity fields from the NetCDF file using the NCDatasets.jl package into arrays then use the

```julia
set!(model, T=some_array, S=some_other_array)
```

function to initialize your model fields using arrays. `set!` also accepts constants and functions of (x, y, z).

**Reply by glwagner** [reply]
A nice way to use `set!` with data might be to use `Interpolations.jl`. You can 

1. Load your data from file into an array;
2. Use `Interpolations.jl` to define a function `interpolated_temperature_data(x, y, z)` that interpolates the discrete data (on whatever grid it's on) to the `z` value being used in the simulation.
 
This will allow you to design a flexible script in which the model resolution is easily changed, and also to use the fancy interpolation features of `Interpolations.jl` if desired.
